### PR TITLE
Fh 3635 - Change the cmds $fhc env * to follow the V3 standard and add unit tests

### PR DESCRIPTION
**Pull request by camilamacedo86**
JIRA: https://issues.jboss.org/browse/FH-3635

Following the changes.

- Move the command for FH3
- Change for all commands to follow the V3 standard 
- Add unit test
- Fix the  `fhc env push` command
- Fix the `fhc env update` command
- Change the URLs of the calls to be the same used by studio ( fh-ngui )
- Remove the command `fhc env listDeplyoed` wich was broke .

**Not found automated tests for this commands**

**Steps to Verify**

```
./bin/fhc.js target
./bin/fhc.js login

./bin/fhc.js env list --app=<cloud-app-id>
./bin/fhc.js env list --app=<cloud-app-id> —json
./bin/fhc.js env create --app=<cloud-app-id>  --name=<nameEnv> --value=<valueEnv>
./bin/fhc.js env create --app=<cloud-app-id>  --name=<nameEnv> --value=<valueEnv> --json
./bin/fhc.js env read --app=<cloud-app-id>  --id=<envId> 
./bin/fhc.js env read --app=<cloud-app-id>  --id=<envId> --json
./bin/fhc.js env unset --app=<cloud-app-id>  --id=<envId> 
./bin/fhc.js env push --app=<cloud-app-id>  
./bin/fhc.js env update --app=<cloud-app-id> --id=<envVarId> --value=<value>
./bin/fhc.js env update --app=<cloud-app-id> --id=<envVarId> --value=<value> --json
./bin/fhc.js env delete --app=<cloud-app-id>  --id=<envId>
```
 
**Local Tests**
* `./bin/fhc.js env list --app=<cloud-app-id>`

screen shot 2017-06-14 at 5 39 13 pm

* `./bin/fhc.js env list --app=<cloud-app-id> --json`

```
Camilas-MacBook-Pro:fh-fhc cmacedo$ ./bin/fhc.js env list --app=3ukgif3bygcn54fd4sysnkkc --json
[{
  "businessObject": "cluster/reseller/customer/domain/project/cloud-apps/environment-variable",
  "domain": "support",
  "varValues": {
    "dev": "valorupdate"
  },
  "varName": "Test2",
  "masked": false,
  "appId": "3ukgif3bygcn54fd4sysnkkc",
  "sysGroupFlags": 65567,
  "sysGroupList": "",
  "sysVersion": 1,
  "sysOwner": "i4ze7ruomd27wnohth22to4v",
  "sysCreated": 1497466411454,
  "guid": "nqnbvq2u3lvmu2xnehvmtpca",
  "sysModified": 1497469701340
},
{
  "businessObject": "cluster/reseller/customer/domain/project/cloud-apps/environment-variable",
  "domain": "support",
  "varValues": {},
  "varName": "Test",
  "masked": false,
  "appId": "3ukgif3bygcn54fd4sysnkkc",
  "sysGroupFlags": 65567,
  "sysGroupList": "",
  "sysVersion": 0,
  "sysOwner": "i4ze7ruomd27wnohth22to4v",
  "sysCreated": 1497466373161,
  "guid": "nrbwq2alymveqpbuj5tlzgtl",
  "sysModified": 1497466373167
}]
Camilas-MacBook-Pro:fh-fhc cmacedo$ 
```
* `./bin/fhc.js env create --app=<cloud-app-id>  --name=<nameEnv> --value=<valueEnv>`

```
Camilas-MacBook-Pro:fh-fhc cmacedo$ ./bin/fhc.js env create --app=3ukgif3bygcn54fd4sysnkkc  --name=TestPR --value=TestV
Environment variables for the app '3ukgif3bygcn54fd4sysnkkc' with the name 'TestPR' and value 'TestV' into the environment 'dev' created successfully.
Camilas-MacBook-Pro:fh-fhc cmacedo$ 
```

* `./bin/fhc.js env create --app=<cloud-app-id>  --name=<nameEnv> --value=<valueEnv> --json`

```
Camilas-MacBook-Pro:fh-fhc cmacedo$ ./bin/fhc.js env create --app=3ukgif3bygcn54fd4sysnkkc  --name=TestPR2 --value=TestV --json
{
  "businessObject": "cluster/reseller/customer/domain/project/cloud-apps/environment-variable",
  "domain": "support",
  "varValues": {
    "dev": "TestV"
  },
  "varName": "TestPR2",
  "masked": false,
  "appId": "3ukgif3bygcn54fd4sysnkkc",
  "sysGroupFlags": 65567,
  "sysGroupList": "",
  "sysVersion": 0,
  "sysOwner": "i4ze7ruomd27wnohth22to4v",
  "sysCreated": 1497472971654,
  "guid": "q7c444d26qirrkejvigyjl4n",
  "sysModified": 1497472971661
}
Camilas-MacBook-Pro:fh-fhc cmacedo$ 
```

* `./bin/fhc.js env read --app=<cloud-app-id>  --id=<envId>` 

screen shot 2017-06-14 at 5 44 53 pm

* `./bin/fhc.js env read --app=<cloud-app-id>  --id=<envId> --json`

```
Camilas-MacBook-Pro:fh-fhc cmacedo$ ./bin/fhc.js env read --app=3ukgif3bygcn54fd4sysnkkc  --id=nqnbvq2u3lvmu2xnehvmtpca --json
{
  "businessObject": "cluster/reseller/customer/domain/project/cloud-apps/environment-variable",
  "domain": "support",
  "varValues": {
    "dev": "valorupdate"
  },
  "varName": "Test2",
  "masked": false,
  "appId": "3ukgif3bygcn54fd4sysnkkc",
  "sysGroupFlags": 65567,
  "sysGroupList": "",
  "sysVersion": 1,
  "sysOwner": "i4ze7ruomd27wnohth22to4v",
  "sysCreated": 1497466411454,
  "guid": "nqnbvq2u3lvmu2xnehvmtpca",
  "sysModified": 1497469701340
}
Camilas-MacBook-Pro:fh-fhc cmacedo$ 
```
* `./bin/fhc.js env delete --app=<cloud-app-id>  --id=<envId> --json`
```
Camilas-MacBook-Pro:fh-fhc cmacedo$ ./bin/fhc.js env delete 3ukgif3bygcn54fd4sysnkkc nqnbvq2u3lvmu2xnehvmtpca
{
  "businessObject": "cluster/reseller/customer/domain/project/cloud-apps/environment-variable",
  "domain": "support",
  "varValues": {},
  "varName": "Test2",
  "masked": false,
  "appId": "3ukgif3bygcn54fd4sysnkkc",
  "sysGroupFlags": 65567,
  "sysGroupList": "",
  "sysVersion": 1,
  "sysOwner": "i4ze7ruomd27wnohth22to4v",
  "sysCreated": 1497466411454,
  "guid": "nqnbvq2u3lvmu2xnehvmtpca",
  "sysModified": 1497469701340
}

```

* `./bin/fhc.js env push --app=<cloud-app-id> `

```
Camilas-MacBook-Pro:fh-fhc cmacedo$ ./bin/fhc.js env push --app=3ukgif3bygcn54fd4sysnkkc 
{
  "message": "",
  "status": "ok",
  "result": null
}
Camilas-MacBook-Pro:fh-fhc cmacedo$ 
```
* `./bin/fhc.js env unset --app=<cloud-app-id>  --id=<envId> `

```
Camilas-MacBook-Pro:fh-fhc cmacedo$ ./bin/fhc.js env unset --app=3ukgif3bygcn54fd4sysnkkc --id=rjgqg453paxp2dcwf77hgn3d
{
  "businessObject": "cluster/reseller/customer/domain/project/cloud-apps/environment-variable",
  "domain": "support",
  "varValues": {},
  "varName": "TestPR",
  "masked": false,
  "appId": "3ukgif3bygcn54fd4sysnkkc",
  "sysGroupFlags": 65567,
  "sysGroupList": "",
  "sysVersion": 1,
  "sysOwner": "i4ze7ruomd27wnohth22to4v",
  "sysCreated": 1497472931708,
  "guid": "rjgqg453paxp2dcwf77hgn3d",
  "sysModified": 1497474491262
}
```
*`./bin/fhc.js env update --app=<cloud-app-id> --id=<envId> --value=<value>

screen shot 2017-06-14 at 6 08 59 pm
`

*`./bin/fhc.js env update --app=<cloud-app-id> --id=<envId> --value=<value> --json`

```
Camilas-MacBook-Pro:fh-fhc cmacedo$ ./bin/fhc.js env update --app=3ukgif3bygcn54fd4sysnkkc --id=rjgqg453paxp2dcwf77hgn3d value=New --json
{
  "businessObject": "cluster/reseller/customer/domain/project/cloud-apps/environment-variable",
  "domain": "support",
  "varName": "TestPR",
  "varValues": {},
  "masked": false,
  "appId": "3ukgif3bygcn54fd4sysnkkc",
  "sysGroupFlags": 65567,
  "sysGroupList": "",
  "sysVersion": 3,
  "sysOwner": "i4ze7ruomd27wnohth22to4v",
  "sysCreated": 1497472931708,
  "guid": "rjgqg453paxp2dcwf77hgn3d",
  "sysModified": 1497474539840
}
Camilas-MacBook-Pro:fh-fhc cmacedo$ 
```

### CODE REVIEW

**Comment by camilamacedo86** [comment]
@david-martin to move these commands of fhc env was required do fix and change the REST URL to use the same that is used into NGUI. Could you check and review this one ?

File: lib/cmd/fh3/env/delete.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/* globals i18n */
+module.exports = {
+  'desc' : i18n._('Deleate cloud app environment variables'),
```
Comment (pb82) [inline lib/cmd/fh3/env/delete.js]: Typo: `Deleate` -> `Delete`

File: lib/cmd/fh3/env/delete.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    'json': 'j',
+    0: 'app',
+    1: 'id',
+    3: 'env',
+    4: 'json'
+  },
+  'describe' : {
+    'app' : i18n._("Unique 24 character GUID of the cloud app."),
+    'id' : i18n._("Unique 24 character GUID of the environment variable."),
+    'env': i18n._("Default value is dev. Environment ID which you want create the environment variable."),
```
Comment (pb82) [inline lib/cmd/fh3/env/delete.js]: missing `for`: `Environment ID for which you want to create the environment variable`

File: lib/cmd/fh3/env/list.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    'app': 'a',
+    'env': 'e',
+    'json': 'j',
+    0: 'app',
+    1: 'env',
+    2: 'json'
+  },
+  'describe' : {
+    'app' : i18n._("Unique 24 character GUID of the cloud app."),
+    'env': i18n._("Default value is dev. Environment ID which you want create the environment variable."),
```
Comment (pb82) [inline lib/cmd/fh3/env/list.js]: see above

File: lib/cmd/fh3/env/create.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+var common = require("../../../common");
+var fhreq = require("../../../utils/request");
+var util = require('util');
+
+module.exports = {
+  'desc' : i18n._('Create cloud app environment variables'),
+  'examples' :
+    [{
+      cmd : 'fhc env create --app=<app> --name=<name> --value=<value> --env=<environment>',
+      desc : "Create a cloud app environment variables for the <app> in the <env> with the <name> and <value>"
```
Comment (tagoh) [inline lib/cmd/fh3/env/create.js]: desc string should be passed to i18n._ to be localized.

File: lib/cmd/fh3/env/push.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  'demand' : ['app'],
+  'alias' : {
+    'app': 'a',
+    'env': 'e',
+    0: 'app',
+    1: 'env'
+  },
+  'describe' : {
+    'app' : i18n._("Unique 24 character GUID of the cloud app."),
+    'env': i18n._("Default value is dev. Environment ID which you want create the environment variable.")
```
Comment (pb82) [inline lib/cmd/fh3/env/push.js]: see above

File: lib/cmd/fh3/env/delete.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/* globals i18n */
+module.exports = {
+  'desc' : i18n._('Deleate cloud app environment variables'),
+  'examples' :
+    [{
+      cmd : 'fhc env delete --app=<app> --id=<id> --env=<environment>',
+      desc : "Delete the cloud app environment variables with the <id> from the <app> in the <env>"
```
Comment (tagoh) [inline lib/cmd/fh3/env/delete.js]: desc string should be passed to i18n._ to be localized.

File: lib/cmd/fh3/env/list.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
+/* globals i18n */
+var common = require("../../../common");
+var fhreq = require("../../../utils/request");
+
+module.exports = {
+  'desc' : i18n._('List cloud app environment variables'),
+  'examples' :
+    [{
+      cmd : 'fhc env list --app=<app> --env=<environment>',
+      desc : "List cloud app environment variables from the <app> and <env>"
```
Comment (tagoh) [inline lib/cmd/fh3/env/list.js]: desc string should be passed to i18n._ to be localized.

File: lib/cmd/fh3/env/push.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/* globals i18n */
+module.exports = {
+  'desc' : i18n._('Push cloud app environment variables'),
+  'examples' :
+    [{
+      cmd : 'fhc env push --app=<app> --env=<environment>',
+      desc : "Push cloud app environment variables from the <app> and <env>"
```
Comment (tagoh) [inline lib/cmd/fh3/env/push.js]: desc string should be passed to i18n._ to be localized.

File: lib/cmd/fh3/env/read.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+var common = require("../../../common");
+var fhc = require("../../../fhc");
+var util = require('util');
+
+module.exports = {
+  'desc' : i18n._('Read cloud app environment variables'),
+  'examples' :
+    [{
+      cmd : 'fhc env read --app=<app> --id=<id> --env=<environment>',
+      desc : "Read a cloud app environment variables with the <id> from the <app> into the <env>"
```
Comment (tagoh) [inline lib/cmd/fh3/env/read.js]: desc string should be passed to i18n._ to be localized.

File: lib/cmd/fh3/env/unset.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
+/* globals i18n */
+module.exports = {
+  'desc' : i18n._('Unset cloud app environment variables'),
+  'examples' :
+    [{
+      cmd : 'fhc env unset --app=<app> --id=<id> --env=<environment>',
+      desc : "Unset cloud app environment variables from the <app> and <env>"
```
Comment (tagoh) [inline lib/cmd/fh3/env/unset.js]: desc string should be passed to i18n._ to be localized.

File: lib/cmd/fh3/env/update.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+var common = require("../../../common");
+var fhreq = require("../../../utils/request");
+var fhc = require("../../../fhc");
+
+module.exports = {
+  'desc' : i18n._('Update cloud app environment variables'),
+  'examples' :
+    [{
+      cmd : 'fhc env update --app=<app> --id=<id> --value=<value> --env=<environment>',
+      desc : "Update the cloud app environment variables with the <id> from the <app> in the <env> with the <name> and <value>"
```
Comment (tagoh) [inline lib/cmd/fh3/env/update.js]: desc string should be passed to i18n._ to be localized.

File: lib/cmd/fh3/env/read.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+      var found = findEnvironment(params,data);
+      if ( found ) {
+        if (!params.json) {
+          params._table = common.createTableForAppEnvVars([found], params.env);
+          return cb(null, params);
+        } else {
+          return cb(null, found);
+        }
+      }
+      return cb(null, util.format(i18n._("Not found environment variable for the app '%s' with the id '%s'"), params.app, params.id));
```
Comment (pb82) [inline lib/cmd/fh3/env/read.js]: Maybe better write: `Unable to find environment variable for app with id ...`

File: lib/cmd/fh3/env/read.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    'id': 'i',
+    'env': 'e',
+    0: 'app',
+    1: 'id',
+    2: 'env'
+  },
+  'describe' : {
+    'app' : i18n._("Unique 24 character GUID of the cloud app."),
+    'id' : i18n._("Unique 24 character GUID of the environment variable."),
+    'env': i18n._("Default value is dev. Environment ID which you want create the environment variable.")
```
Comment (pb82) [inline lib/cmd/fh3/env/read.js]: see above

File: lib/cmd/fh3/env/unset.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    'json': 'j',
+    0: 'app',
+    1: 'id',
+    2: 'env',
+    3: 'json'
+  },
+  'describe' : {
+    'app' : i18n._("Unique 24 character GUID of the cloud app."),
+    'id' : i18n._("Unique 24 character GUID of the environment variable."),
+    'env': i18n._("Default value is dev. Environment ID which you want create the environment variable."),
```
Comment (pb82) [inline lib/cmd/fh3/env/unset.js]: see above

**Comment by david-martin** [comment]
@camilamacedo86 I get this error with the list cmd

```
fhc env list --app=nb626ziawyu6e45mdluolo2p
fhc ERR! cb() never called!
fhc ERR!  
fhc Command executed with error
/home/dmartin/work/fh-fhc/node_modules/cli-table/lib/cli-table/index.js:93
        var width = typeof cell == 'object' && cell.width != undefined
                                                   ^

TypeError: Cannot read property 'width' of null
    at /home/dmartin/work/fh-fhc/node_modules/cli-table/lib/cli-table/index.js:93:52
    at Array.forEach (native)
    at /home/dmartin/work/fh-fhc/node_modules/cli-table/lib/cli-table/index.js:92:13
    at Array.forEach (native)
    at Table.toString (/home/dmartin/work/fh-fhc/node_modules/cli-table/lib/cli-table/index.js:91:34)
    at Object.handler.printAsTable (/home/dmartin/work/fh-fhc/lib/utils/result-handler.js:43:23)
    at Object.handler.printFormatted (/home/dmartin/work/fh-fhc/lib/utils/result-handler.js:23:12)
    at Object.handler.printCommandResult (/home/dmartin/work/fh-fhc/lib/utils/result-handler.js:16:14)
    at /home/dmartin/work/fh-fhc/bin/fhc.js:18:21
    at /home/dmartin/work/fh-fhc/lib/cmd/fh3/env/list.js:35:16
```

The `--json` arg works ok
```
fhc env list --app=nb626ziawyu6e45mdluolo2p --json
[{
  "businessObject": "cluster/reseller/customer/domain/project/cloud-apps/environment-variable",
  "domain": "testing",
  "appId": "nb626ziawyu6e45mdluolo2p",
  "varName": "FH_SKIP_MBAAS_AUTH",
  "varValues": {
    "dev": true,
    "test": true,
    "live": true
  },
  "masked": false,
  "sysOwner": "mivux355rlqsugyp5kimmaic",
  "guid": null,
  "sysModified": 1498819329646,
  "sysVersion": 0,
  "sysCreated": 1498819329646,
  "sysGroupFlags": 70110,
  "sysGroupList": ""
}]
```

I'm trying to track it down, but maybe you have some idea?

File: lib/cmd/fh3/env/read.js (hunk 12 above)
Comment (pb82) [inline lib/cmd/fh3/env/read.js]: Shouldn't it return an error if the environment is not found? (`return cb(new Error(i18n._("Not found environment variable for the app '%s' with the id '%s'"), params.app, params.id)))`)

File: lib/cmd/fh3/env/update.js [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    'json' : i18n._("Output in json format")
+  },
+  'customCmd' : function(params,cb) {
+    params.env = params.env || 'dev';
+    fhc.env.read({app:params.app, id:params.id, env:params.env, json:true}, function(err,data) {
+      if (err) {
+        return cb(err);
+      }
+      var url = "/box/api/apps/" + params.app + "/env/" + params.env + "/envvars/" + params.id;
+      data.value = params.value;
```
Comment (pb82) [inline lib/cmd/fh3/env/update.js]: Data can be `undefined` if the `read` command doesn't find the env. You need to either handle that case here or (better) make the read command error when the env is not found. What do you think?

File: lib/cmd/fh3/env/create.js [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  'customCmd' : function(params, cb) {
+    var environment = params.env || 'dev';
+    var url = "/box/api/apps/" + params.app + "/env/" + environment + "/envvars";
+    var playload = { masked: "false", name: params.name, value: params.value};
+    return common.doApiCall(fhreq.getFeedHenryUrl(), url, playload, i18n._("Error creating env var: "), function(err, data) {
+      if (err) {
+        return cb(err);
+      }
+      if (!params.json) {
+        return cb(null, util.format(i18n._("Environment variables for the app '%s' with the name '%s' and value '%s' into the environment '%s' created successfully."),params.app,params.name, params.value, environment));
```
Comment (david-martin) [inline lib/cmd/fh3/env/create.js]: This wording is a little tough to read.

Might I suggest something like this instead
*Successfully created Environment variable for the app '%s' in the '%s' Environment: '%s=%s'*

e.g. 
*Successfully created Environment variable for the app 'nb626ziawyu6e45mdluolo2p' in the 'dev' Environment: 'TEST_ENV_VAR=testvalue'*

**Comment by david-martin** [comment]
@camilamacedo86 I've gone through the various arguments, and they seem to be working OK.

The only thing is the example you gave for updating an env var.
I had to use this:
```
./bin/fhc.js env update --app=<cloud-app-id> --id=<envVarId> --value=<value>
./bin/fhc.js env update --app=<cloud-app-id> --id=<envVarId> --value=<value> --json
```
instead of this:
```
./bin/fhc.js env update --app=<cloud-app-id> --value=<valueEnv> --value=<value>
./bin/fhc.js env update --app=<cloud-app-id> --value=<valueEnv> --value=<value> --json
```


e.g.
```
fhc env update --app=nb626ziawyu6e45mdluolo2p --id=qfiqqh23egobn64p7pjijwp4 --value=test5 --json
```

**Comment by camilamacedo86** [comment]
@david-martin, tks for the test. I wrote wrong the command in the first comment. I already update it. You test correctly.

File: lib/cmd/fh3/env/delete.js (hunk 1 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/env/delete.js]: Tks for the review 👍 Done.

File: lib/cmd/fh3/env/delete.js (hunk 2 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/env/delete.js]: Tks for the review. Done 👍

File: lib/cmd/fh3/env/list.js (hunk 3 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/env/list.js]: Tks. Done 👍

File: lib/cmd/fh3/env/push.js (hunk 5 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/env/push.js]: Tks. Done 👍

File: lib/cmd/fh3/env/read.js (hunk 12 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/env/read.js]: Tks Done 👍

File: lib/cmd/fh3/env/read.js (hunk 13 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/env/read.js]: Done 👍 Tks

File: lib/cmd/fh3/env/unset.js (hunk 14 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/env/unset.js]: Done 👍

File: lib/cmd/fh3/env/create.js (hunk 16 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/env/create.js]: Tks for the review. Done 👍

File: lib/cmd/fh3/env/update.js (hunk 15 above)
Comment (camilamacedo86) [inline lib/cmd/fh3/env/update.js]: Tks. Done 👍

**Comment by camilamacedo86** [comment]
Hi, @david-martin @pb82 @tagoh , really tks for the review. All changes required are done. 
I am looking for the approval to merge it.

**Pull request merged**